src/app/server/CommissioningWindowManager.cpp

```diff
@@ -547,6 +552,23 @@ void CommissioningWindowManager::UpdateWindowStatus(CommissioningWindowStatusEnu
         MatterReportingAttributeChangeCallback(kRootEndpointId, AdministratorCommissioning::Id,
                                                AdministratorCommissioning::Attributes::WindowStatus::Id);
     }
+
+#if CHIP_DEVICE_CONFIG_ENABLE_THREAD && CHIP_DEVICE_CONFIG_THREAD_FTD
+    // Block device role changing into Router if commissioning window opened and device not yet Router.
+    if (mWindowStatus == CommissioningWindowStatusEnum::kEnhancedWindowOpen &&
```

Comment:
1. What if basic commissioning window is opened?
2. OpenThread APIs shouldn't be used in a common component since there are other Thread implementations, such as Tizen, that may want to use the same logic.
3. OpenThread functions shouldn't be called from the CHIP thread without the Thread stack lock held. Also, enabling/disabling the router promotion should be atomic to make sure that when `otThreadSetRouterEligible(DeviceLayer::ThreadStackMgrImpl().OTInstance(), false);` is called the device is indeed not a router yet.

I suggest adding an API to ThreadStackManager, such as `SetRouterPromotion(bool)` or `SuppressRouterPromotion(bool)` that would abstract this logic.

Reply:
+1

Reply:
We must disable promotion whenever any commissioning window is open, not just enhanced